docs/source/extending_pywr/extending_pywr_parameters.rst

```diff
@@ -0,0 +1,224 @@
+
+To get the value of a child parameter use the ``Parameter.get_value`` method, or for the index use
+``Parameter.get_index``. These methods return the value/index for the current timestep and scenario. To access the
+value from previous timesteps you must manually track the state of the child parameters.
+
+The ``pywr.parameters.load_parameter`` function is used to load parameters from JSON. This works with both references to
+parameters and nested parameters.
+
+As an example, see a simplified version of ``AggregatedParameter`` that returns the sum value of it's child parameters.
+
+.. code-block:: python
+
+    class SumParameter(Parameter):
+        def __init__(self, model, parameters):
+            self.parameters = parameters
+            for parameter in self.parameters:
+                self.children.add(parameter)
+
+        def value(self, timestep, scenario_index):
+            total_value = sum([parameter.get_value() for parameter in parameters])
+            return total_value
+
+        @classmethod
+        def load(self, model, data):
+            parameters = [load_parameter(parameter_data)
+                          for parameter_data in data.pop("parameters")]
+            return cls(model, parameters)
+
+
+Improving performance with Cython
+=================================
+
+Parameters are evaluated many times and can be a significant part of the model run time. Many of the parameters in the
+core library have been written in Cython to improve performance. Custom parameters can be written in Cython too. Cython
+can also be used to link to external C/C++ libraries.
+
+A full tutorial in Cython is beyond the scope of this documentation - see the
+`Cython Documentation <https://cython.readthedocs.io/en/latest/>`_.
+
+The easiest way to compile and run custom parameters written in Cython is using the ``pyximport`` command, which
+compiles pyx modules at runtime. If you parameter is linking to a foreign library you may need to compile using a
+``setup.py`` in order to pass linker arguments.
+
+The example below demonstrates a custom parameter which uses a function from a foreign library (the ``pow`` function
+from ``libm``). There are a few differences from the Python equivalent:
+
+* Use of the ``cimport`` statement
+* Inherit from ``pywr.parameters._parameters.Parameter``
+* The ``value`` method is defined as a cpdef function. This signature must match exactly.
+* There is no benefit to using cdef/cpdef for the ``__init__`` and ``load`` methods as they are only called once and
+  are not performance critical.
+
+.. code-block:: cython
+    :caption: custom_parameters.pyx
+
+    from pywr.parameters._parameters cimport Parameter
+    from pywr._core cimport Timestep, ScenarioIndex
+
+    cdef extern from "math.h":
+        double pow(double x, double y)
+
+    cdef class SquaredParameter(Parameter):
+        cdef double _value
+
+        def __init__(self, model, value):
+            super().__init__(model)
```

Comment:
Does `super()` (no arguments) work in cython?

Reply:
Yes. This was the only example I actually tested.